src/pool-bin/libraries/BinPool.sol

```diff
@@ -182,50 +189,64 @@ library BinPool {
         uint16 binStep;
     }
 
+    struct SwapState {
+        uint24 activeId;
+        uint24 protocolFee;
+        uint24 swapFee;
+        bytes32 feeForProtocol;
+    }
+
     function swap(State storage self, SwapParams memory params, uint128 amountIn)
         internal
-        returns (BalanceDelta result, bytes32 feeForProtocol, uint24 activeId, uint24 swapFee)
+        returns (BalanceDelta result, SwapState memory swapState)
     {
-        activeId = self.slot0.activeId;
+        Slot0 memory slot0Cache = self.slot0;
+        swapState.activeId = slot0Cache.activeId;
+        swapState.protocolFee =
+            params.swapForY ? slot0Cache.protocolFee.getOneForZeroFee() : slot0Cache.protocolFee.getZeroForOneFee();
 
         if (amountIn == 0) revert BinPool__InsufficientAmountIn();
 
         bytes32 amountsLeft = params.swapForY ? amountIn.encodeFirst() : amountIn.encodeSecond();
         bytes32 amountsOut;
-        swapFee = self.slot0.swapFee;
+
+        /// @dev swap fee includes protocolFee (charged first) and lpFee
+        swapState.swapFee = swapState.protocolFee.calculateSwapFee(slot0Cache.lpFee);
 
         while (true) {
-            bytes32 binReserves = self.reserveOfBin[activeId];
+            bytes32 binReserves = self.reserveOfBin[swapState.activeId];
             if (!binReserves.isEmpty(!params.swapForY)) {
-                (bytes32 amountsInWithFees, bytes32 amountsOutOfBin, bytes32 totalFees) =
-                    binReserves.getAmounts(swapFee, params.binStep, params.swapForY, activeId, amountsLeft);
+                (bytes32 amountsInWithFees, bytes32 amountsOutOfBin,) = binReserves.getAmounts(
+                    swapState.swapFee, params.binStep, params.swapForY, swapState.activeId, amountsLeft
+                );
 
                 if (amountsInWithFees > 0) {
                     amountsLeft = amountsLeft.sub(amountsInWithFees);
                     amountsOut = amountsOut.add(amountsOutOfBin);
 
-                    bytes32 pFee = totalFees.getExternalFeeAmt(self.slot0.protocolFee);
+                    /// @dev calc protocol fee for current bin, charged on amoutIn and charged before lpFee
+                    bytes32 pFee = amountsInWithFees.getExternalFeeAmt(slot0Cache.protocolFee);
```

Comment:
instead of 
`totalFees * protocolFeePercent` : this implies protocol fee is part of swap fee

now we have

`amountsInWithFees  * protocolFeePercent`: this implies protocol fee is charged directly on input amount